modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4SizeHeaderFrameDecoder.java

```diff
@@ -37,17 +37,17 @@
     protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
         try {
             BytesReference networkBytes = Netty4Utils.toBytesReference(in);
-            int messageLength = TcpTransport.readMessageLength(networkBytes) + HEADER_SIZE;
+            int messageLength = TcpTransport.readMessageLength(networkBytes);
+            int messageWithHeaderLength = messageLength + HEADER_SIZE;
             // If the message length is -1, we have not read a complete header. If the message length is
             // greater than the network bytes available, we have not read a complete frame.
-            if (messageLength != -1 && messageLength <= networkBytes.length()) {
+            if (messageLength != -1 && messageWithHeaderLength <= networkBytes.length()) {
                 final ByteBuf message = in.skipBytes(HEADER_SIZE);
                 // 6 bytes would mean it is a ping. And we should ignore.
-                if (messageLength != 6) {
+                if (messageWithHeaderLength != 6) {
```

Comment:
I think it is more clear if we just let TcpTransport.readMessageLength() return -1 in case of a ping (right now it returns 0, but TcpTransport.PING_DATA_SIZE itself is -1)